src/lib/optimizer/strategy/nested_expression_rule.cpp

```diff
@@ -0,0 +1,163 @@
+#include "logical_query_plan/abstract_lqp_node.hpp"
+#include "logical_query_plan/lqp_column_reference.hpp"
+#include "logical_query_plan/lqp_expression.hpp"
+#include "logical_query_plan/predicate_node.hpp"
+#include "logical_query_plan/projection_node.hpp"
+#include "resolve_type.hpp"
+#include "utils/arithmetic_operator_expression.hpp"
+
+namespace opossum {
+
+std::string NestedExpressionRule::name() const { return "Nested Expression Rule"; }
+
+bool NestedExpressionRule::apply_to(const std::shared_ptr<AbstractLQPNode>& node) {
+  if (node->type() != LQPNodeType::Projection) {
+    return _apply_to_children(node);
+  }
+
+  auto projection_node = std::dynamic_pointer_cast<ProjectionNode>(node);
+
+  const auto column_expressions = projection_node->column_expressions();
+  for (auto column_id = ColumnID{0}; column_id < column_expressions.size(); column_id++) {
+    const auto expression = column_expressions[column_id];
+    if (!expression->is_arithmetic_operator()) {
+      continue;
+    }
+
+    // If the expression contains operands with different types, or one of the operands is a column reference,
+    // it can't be resolved in this rule, and will be handled by the Projection later.
+    const auto expression_type = _get_type_of_expression(expression);
+    if (expression_type == std::nullopt) {
+      continue;
+    }
+
+    auto value = NULL_VALUE;
+    resolve_data_type(*expression_type, [&](auto type) { value = _evaluate_expression(type, expression); });
+
+    // If the value is still NULL_VALUE, then the expression could not be resolved at this point,
+    // e.g. because it is not an arithmetic operator.
+    if (!variant_is_null(value) &&
+        _replace_expression_in_parents(node, node->output_column_references()[column_id], value)) {
+      // If we successfully replaced the occurences of the Expression in the parent tree
+      // we can remove the column here.
+      _remove_column_from_projection(projection_node, column_id);
+    }
+  }
+
+  return _apply_to_children(node);
+}
+
+bool NestedExpressionRule::_replace_expression_in_parents(const std::shared_ptr<AbstractLQPNode>& node,
+                                                          const LQPColumnReference& expression_column,
+                                                          const AllTypeVariant& value) {
+  auto parent_tree_changed = false;
+  for (auto parent : node->parents()) {
+    if (parent->type() != LQPNodeType::Predicate) {
+      parent_tree_changed |= _replace_expression_in_parents(parent, expression_column, value);
+      continue;
+    }
+
+    auto predicate_node = std::dynamic_pointer_cast<PredicateNode>(parent);
+
+    // We look for a PredicateNode which has an LQPColumnReference as value,
+    // referring to the column which contains the Expression we resolved before.
+    if (is_lqp_column_reference(predicate_node->value()) &&
+        boost::get<LQPColumnReference>(predicate_node->value()) == expression_column) {
```

Comment:
`std::get`?

Reply:
`std::get` doesn't seem to work with opossum with our setup here, or do I have to use that differently?
```error: no matching function for call to ‘std::get<opossum::LQPColumnReference>(const AllParameterVariant&)’```

Reply:
Oups